Security/TwoFactor/Provider/TwoFactorProviderRegistry.php

```diff
@@ -70,7 +88,14 @@ public function requestAuthenticationCode(AuthenticationContextInterface $contex
 
                 // Set authentication completed
                 if ($context->isAuthenticated()) {
+                    if (null !== $this->eventDispatcher) {
+                        $this->eventDispatcher->dispatch(TwoFactorAuthEvent::NAME, new TwoFactorAuthEvent());
+                    }
                     $this->flagManager->setComplete($providerName, $token);
+                } else {
+                    if (null !== $this->eventDispatcher && $context->getRequest()->request->has($this->authRequestParameter)) {
```

Comment:
No need to check $this->eventDispatcher for null when constructor injection is used.